File: ts/src/core/citations.ts
/**
 * Ge'ez numeral and citation parsing, plus Ethiopic text folding for search.
 *
 * The folding maps every homophone fidel series to one canonical series so
 * that spellings which sound identical compare equal: ሐ/ኀ fold into ሀ,
 * ሠ into ሰ, ዐ into አ, ፀ into ጸ, and the interchangeable first/fourth
 * orders of the guttural series fold together (ሀ/ሃ, አ/ኣ). Arabic digits
 * fold into Ge'ez numerals so "2 ጴጥሮስ" matches "፪ኛ ጴጥሮስ".
 *
 * The citation grammar accepts the printed Gitsawe style (ም· ፲ ቍ· ፩ – ፳፪),
 * plain chapter:verse ranges (3:16-18), and space-separated numerals.
 * The Python twin is py/eotc/citations.py; shared fixtures hold them equal.
 */

const GEEZ_DIGITS = new Map([
  ['፩', 1], ['፪', 2], ['፫', 3], ['፬', 4], ['፭', 5],
  ['፮', 6], ['፯', 7], ['፰', 8], ['፱', 9], ['፲', 10],
  ['፳', 20], ['፴', 30], ['፵', 40], ['፶', 50], ['፷', 60],
  ['፸', 70], ['፹', 80], ['፺', 90], ['፻', 100], ['፼', 10000],
]);

export function geezToInteger(raw: string): number | null {
  let total = 0, group = 0, run = 0;
  let found = false;
  for (const char of raw || '') {
    const value = GEEZ_DIGITS.get(char);
    if (!value) continue;
    found = true;
    if (value === 100) { group += (run || 1) * 100; run = 0; }
    else if (value === 10000) { total += (group + run || 1) * 10000; group = 0; run = 0; }
    else run += value;
  }
  return found ? total + group + run : null;
}

// Homophone series: each string pairs source characters with the canonical
// character at the same vowel order. Kept flat so the Python twin can share
// the exact table.
const FOLD_PAIRS: [string, string][] = [
  ['ሐሑሒሓሔሕሖ', 'ሀሁሂሃሄህሆ'],
  ['ኀኁኂኃኄኅኆ', 'ሀሁሂሃሄህሆ'],
  ['ኻኺኼኽ', 'ሃሂሄህ'],
  ['ሠሡሢሣሤሥሦ', 'ሰሱሲሳሴስሶ'],
  ['ዐዑዒዓዔዕዖ', 'አኡኢኣኤእኦ'],
  ['ፀፁፂፃፄፅፆ', 'ጸጹጺጻጼጽጾ'],
  ['1234567890', '፩፪፫፬፭፮፯፰፱0'],
];
const FOLD_MAP = new Map<string, string>();
for (const [from, to] of FOLD_PAIRS) {
  for (let i = 0; i < from.length; i++) FOLD_MAP.set(from[i], to[i]);
}
// The first and fourth orders of the guttural series are used interchangeably.
FOLD_MAP.set('ሃ', 'ሀ');
FOLD_MAP.set('ኣ', 'አ');

/**
 * Fold Ethiopic text for comparison: homophone series unified, punctuation
 * and whitespace removed, Latin lowercased, Arabic digits as Ge'ez numerals.
 */
export function foldEthiopic(raw: string): string {
  let out = '';
  for (const char of (raw || '').normalize('NFC').toLowerCase()) {
    if (/[\s፡።፣፤፥፦፧·‧.,:;!?"'“”‘’()\[\]{}\-–—/\\]/.test(char)) continue;
    out += FOLD_MAP.get(char) ?? char;
  }
  return out;
}

export interface ParsedCitation {
  chapter: number | null;
  verseStart: number | null;
  verseEnd: number | null;
  toEndOfChapter: boolean;
}

/**
 * Parse the numeric part of a citation. Handles the printed Gitsawe grammar
 * (chapter after ም, verses after ቍ, ፍጻሜ ምዕራፍ as "to end of chapter"),
 * plain `3:16-18`, and bare space-separated numbers (`3 16 18`).
 */
export function parseCitation(raw: string): ParsedCitation {
  const source = (raw || '').trim();
  const none: ParsedCitation = { chapter: null, verseStart: null, verseEnd: null, toEndOfChapter: false };
  if (!source) return none;

  const latin = /^(\d+)(?:\s*[:.]\s*(\d+)(?:\s*[-–]\s*(\d+))?)?$/.exec(source);
  if (latin) {
    return {
      chapter: Number(latin[1]),
      verseStart: latin[2] ? Number(latin[2]) : null,
      verseEnd: latin[3] ? Number(latin[3]) : null,
      toEndOfChapter: false,
    };
  }

  if (/[ቍቊ]/.test(source) || /ም/.test(source) || /[፩-፼]/.test(source)) {
    const parts = source.split(/[ቍቊ]/);
    const left = parts[0] ?? '';
    const right = parts.slice(1).join(' ');
    const toEndOfChapter = /ፍ\s*[፡፣።፥፤,.:·]?\s*ም/.test(source);
    const chapterMatch = left.replace(/^\s*ም(?:ዕራፍ)?\s*[·.:፡]?/, '').match(/[፩-፼]+|\d+/);
    const readNumber = (token: string) => /^\d+$/.test(token) ? Number(token) : geezToInteger(token);
    let verseValues: number[] = [];
    if (parts.length > 1) {
      verseValues = [...right.matchAll(/[፩-፼]+|\d+/g)]
        .map((match) => readNumber(match[0])).filter((n): n is number => n !== null && n > 0);
    } else {
      // No ቍ marker: first number is the chapter, the rest are verses.
      const all = [...left.matchAll(/[፩-፼]+|\d+/g)]
        .map((match) => readNumber(match[0])).filter((n): n is number => n !== null && n > 0);
      verseValues = all.slice(1);
      if (all.length) return {
        chapter: all[0],
        verseStart: verseValues[0] ?? null,
        verseEnd: toEndOfChapter ? null : (verseValues.length > 1 ? verseValues[verseValues.length - 1] : null),
        toEndOfChapter,
      };
      return { ...none, toEndOfChapter };
    }
    return {
      chapter: chapterMatch ? readNumber(chapterMatch[0]) : null,
      verseStart: verseValues[0] ?? null,
      verseEnd: toEndOfChapter ? null : (verseValues.length > 1 ? verseValues[verseValues.length - 1] : null),
      toEndOfChapter,
    };
  }

  return none;
}

File: py/eotc/citations.py
"""
Ge'ez numeral and citation parsing, plus Ethiopic text folding for search.

The TypeScript twin is ts/src/core/citations.ts; shared response fixtures
hold the two implementations byte-identical. Keep the tables and the parsing
rules in lockstep.
"""

from __future__ import annotations

import re

_GEEZ_DIGITS = {
    "፩": 1, "፪": 2, "፫": 3, "፬": 4, "፭": 5,
    "፮": 6, "፯": 7, "፰": 8, "፱": 9, "፲": 10,
    "፳": 20, "፴": 30, "፵": 40, "፶": 50, "፷": 60,
    "፸": 70, "፹": 80, "፺": 90, "፻": 100, "፼": 10000,
}


def geez_to_integer(raw: str) -> int | None:
    total = group = run = 0
    found = False
    for char in raw or "":
        value = _GEEZ_DIGITS.get(char)
        if not value:
            continue
        found = True
        if value == 100:
            group += (run or 1) * 100
            run = 0
        elif value == 10000:
            total += (group + run or 1) * 10000
            group = run = 0
        else:
            run += value
    return total + group + run if found else None


_FOLD_PAIRS = [
    ("ሐሑሒሓሔሕሖ", "ሀሁሂሃሄህሆ"),
    ("ኀኁኂኃኄኅኆ", "ሀሁሂሃሄህሆ"),
    ("ኻኺኼኽ", "ሃሂሄህ"),
    ("ሠሡሢሣሤሥሦ", "ሰሱሲሳሴስሶ"),
    ("ዐዑዒዓዔዕዖ", "አኡኢኣኤእኦ"),
    ("ፀፁፂፃፄፅፆ", "ጸጹጺጻጼጽጾ"),
    ("1234567890", "፩፪፫፬፭፮፯፰፱0"),
]
_FOLD_MAP = {src: dst for group, target in _FOLD_PAIRS for src, dst in zip(group, target)}
_FOLD_MAP["ሃ"] = "ሀ"
_FOLD_MAP["ኣ"] = "አ"

_STRIP = set(" \t\n\r፡።፣፤፥፦፧·‧.,:;!?\"'“”‘’()[]{}-–—/\\")


def fold_ethiopic(raw: str) -> str:
    """Fold Ethiopic homophones and strip punctuation for comparison."""
    import unicodedata

    out = []
    for char in unicodedata.normalize("NFC", (raw or "")).lower():
        if char in _STRIP:
            continue
        out.append(_FOLD_MAP.get(char, char))
    return "".join(out)


_LATIN = re.compile(r"^(\d+)(?:\s*[:.]\s*(\d+)(?:\s*[-–]\s*(\d+))?)?$")
_CHAPTER_PREFIX = re.compile(r"^\s*ም(?:ዕራፍ)?\s*[·.:፡]?")
_NUMBER = re.compile(r"[፩-፼]+|\d+")
_TO_END = re.compile(r"ፍ\s*[፡፣።፥፤,.:·]?\s*ም")


def _read_number(token: str) -> int | None:
    # str.isdigit alone is true for Ge'ez numerals too (Unicode digit
    # property), so require ASCII before treating the token as Arabic digits.
    return int(token) if token.isascii() and token.isdigit() else geez_to_integer(token)


def parse_citation(raw: str) -> dict:
    """Parse the numeric part of a citation; mirrors the TypeScript grammar."""
    source = (raw or "").strip()
    none = {"chapter": None, "verseStart": None, "verseEnd": None, "toEndOfChapter": False}
    if not source:
        return dict(none)

    latin = _LATIN.match(source)
    if latin:
        return {
            "chapter": int(latin.group(1)),
            "verseStart": int(latin.group(2)) if latin.group(2) else None,
            "verseEnd": int(latin.group(3)) if latin.group(3) else None,
            "toEndOfChapter": False,
        }

    if re.search(r"[ቍቊ]", source) or "ም" in source or re.search(r"[፩-፼]", source):
        parts = re.split(r"[ቍቊ]", source)
        left = parts[0] if parts else ""
        right = " ".join(parts[1:])
        to_end = bool(_TO_END.search(source))
        chapter_match = _NUMBER.search(_CHAPTER_PREFIX.sub("", left))
        if len(parts) > 1:
            verse_values = [n for n in (_read_number(m.group(0)) for m in _NUMBER.finditer(right))
                            if n is not None and n > 0]
        else:
            all_numbers = [n for n in (_read_number(m.group(0)) for m in _NUMBER.finditer(left))
                           if n is not None and n > 0]
            verse_values = all_numbers[1:]
            if all_numbers:
                return {
                    "chapter": all_numbers[0],
                    "verseStart": verse_values[0] if verse_values else None,
                    "verseEnd": None if to_end else (verse_values[-1] if len(verse_values) > 1 else None),
                    "toEndOfChapter": to_end,
                }
            return {**none, "toEndOfChapter": to_end}
        return {
            "chapter": _read_number(chapter_match.group(0)) if chapter_match else None,
            "verseStart": verse_values[0] if verse_values else None,
            "verseEnd": None if to_end else (verse_values[-1] if len(verse_values) > 1 else None),
            "toEndOfChapter": to_end,
        }

    return dict(none)
